src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -250,7 +250,11 @@ private static Binder GetEnclosingBinder(CSharpSyntaxNode node, int position, Bi
                 {
                     binder = rootBinder.GetBinder(current);
                 }
-                else if (current is ExpressionSyntax && (current.Parent as LambdaExpressionSyntax)?.Body == current)
+                else if (current is ExpressionSyntax && 
+                            ((current.Parent as LambdaExpressionSyntax)?.Body == current ||
+                             (current.Parent as SwitchStatementSyntax)?.Expression == current ||
+                             (current.Parent as ForEachStatementSyntax)?.Expression == current ||
+                             (current.Parent as IfStatementSyntax)?.Condition == current))
```

Comment:
There is no such thing as a match statement. If you are talking about match expression case sections, they are handled in this function.